jax/api.py

```diff
@@ -64,7 +64,7 @@
                   "Disable JIT compilation and just call original Python.")
 
 
-def jit(fun, static_argnums=()):
+def jit(fun, static_argnums=(), device_values=True):
```

Comment:
I think we should be careful about adding keyword arguments like this to the core APIs. It makes the device_get implementation nice, but from an API perspective, it seems odd to have this option when you can call device_get on the outputs. I'm not sure why we have the flag to toggle this behavior either, so maybe I'm missing something, but this doesn't seem useful enough to warrant being part of the jit API.

Reply:
I agree we should be careful, but this is a clear-cut case: it's exactly analogous to how we handle `device_put` on the input side.

Saying we can call `device_get` on outputs is a bit circular, since there is no other `device_get` than this. In other words, this isn't redundant with some other code; quite the opposite, this is the only way to implement `device_get`.

The "`device_get`" on master is implemented differently, but (1) it couldn't handle JaxTuples, ShardedDeviceArrays, or now DeviceTuples or ShardedDeviceTuples (it owuld raise an error), and (2) if it were ever used in a tracing context (e.g. called in a function being differentiated, or in a jitted function) it would raise an error. So in effect `device_get` hasn't been correctly implemented until now.

This new implementation handles all those cases, and with a net reduction in code.

It's pretty clear that this is the right implementation mechanism. Whether we should add the `device_values` argument to the user-facing API, or make it private, is a separate question, and the answer is less clear. But I think it's still better to make it public than keep it private. And it's a Type 2 decision: if we find a clear reason not to support this flag, we can remove its functionality and raise a deprecation warning easily.

Reply:
Would you recommend just adding an underscore, like `_device_values`?

Reply:
I'd recommend renaming jit to_jit_impl or similar, putting the device_values arg there, and have jit call _jit_impl(device_values=True).

Reply:
Got it. I'll do that so we can decouple the API question entirely, and not have to make any API change here.

Great idea! Thanks for your watchful eye defending our API borders.